### PR TITLE
Add FIND server as install option

### PR DESCRIPTION
**Pull request by mstormi**
Framework for Internal Navigation and Discovery
see http://www.internalpositioning.com

Signed-off-by: Markus Storm <markus.storm@gmx.net> (github: mstormi)

### CODE REVIEW

**Comment by kubawolanin** [comment]
Hey Markus, thanks for your contribution!
Looking forward to try it out.

Since we're in a somewhat controlled environment, I wonder if we could add some preconfigured files to `$openhabconfig`?
It would be great if there was some reference to FIND in the docs as well. 
So far I've skimmed through [this topic](https://community.openhab.org/t/find-high-precision-indoor-positioning-framework-for-most-wifi-enabled-devices/9909/37) but some simple "Getting Started" guide for OH users would be beneficial too ;-)

Cheers

**Comment by mstormi** [comment]
Hi Kuba,
those docs are [here](https://github.com/openhab/openhab1-addons/wiki/FIND). I suggest @ThomDietrich moves them to a better place, too, when he's ready to merge this PR.
What sort of preconfigured files do you think of ? I believe it's actually just the item you need to configure to listen to the proper MQTT channel.

**Comment by kubawolanin** [comment]
@mstormi thanks for pointing me the document! I thought the wiki is dead already and [openhab-docs](https://github.com/openhab/openhab-docs/) successfuly replaced it :laughing: 
Yet it's still actively updated and I agree that we should move those tutorials.

If it's just an Item and some configs to be placed in `mqtt.cfg` then maybe we could display an URL to the docs after successful installation? It would make it easier to get started and definitely decrease number of questions on the forum ;-)

Warm regards

**Comment by ThomDietrich** [comment]
I was surprised by this too. The wiki IS supposed to be dead and I'd prefer if people stopped adding content there. The wiki should only be a reference for those out there still using openHAB 1.x.

As with all [Tutorials](http://docs.openhab.org/tutorials/) we've discussed for a few times that the best place to post them and keep them updated is the community. @mstormi would you create a copy of the wiki article there?

As for the configuration: I believe @mstormi's idea was to automatically create a `$OPENHAB_CONF/items/find.items` file with the item in there (plus some helpful comments). I was thinking about that kind of automatic integration for other optional components as well and would agree to do that.

**Comment by mstormi** [comment]
Well that OH1 wiki entry has been there for a long time (I didn't mean to maintain it there, just updated it today when I discovered some typos). Alright, will move that into tutorials on the community site.
I agree it would make sense to show the user a pointer to the community tutorial after he has successfully installed this in openHABian, but I'd rather suggest to put all such links into a big doc and show that whenever someone has selected to install any of the optional openHABian components.

On a sidenote, @ThomDietrich did you browse the 'Application Integration' and 'Misc' links you get to see on the bottom right of the old wiki ? I suggest you get in touch with the author of each and every of those entries and ask him to move his/her entry over to the community as well (or tell him/her where).
I myself have got at least one more article over there (the one how to run that 'cat') but noone ever asked me to move that until you did today. Will move that over to tutorials, too, but we need to trigger the other authors.
PS: back from my ventures  i.e. you can get me a call to discuss that issue #128.

File: openhabian-setup.sh [hunk 1]
```diff
@@ -1033,6 +1033,103 @@ To continue your integration in openHAB 2, please follow the instructions under:
   fi
 }
 
+find_setup() {
+  FAILED=0
+  introtext="This will install and setup the FIND server system to allow for indoor localization of WiFi devices. See further information at
+ http://www.internalpositioning.com/"
+  failtext="Sadly there was a problem setting up the selected option. Please report this problem in the openHAB community forum or as a open
+HABian GitHub issue."
```
Comment (ThomDietrich) [inline openhabian-setup.sh]: Please fix these linebreaks back to single lines.

File: openhabian-setup.sh [hunk 2]
```diff
@@ -1033,6 +1033,103 @@ To continue your integration in openHAB 2, please follow the instructions under:
...
+  cat>$FIND_DEFAULT <<EOF
+#
+# FIND server service configuration
+# see https://www.internalpositioning.com/server/
+#
+ADDRESS=localhost
+PORT=8003
+MQTTSERVER=localhost
+MQTTPORT=1883
+EOF
```
Comment (ThomDietrich) [inline openhabian-setup.sh]: Please move these into files in the includes directory.

**Comment by ThomDietrich** [comment]
Just an idea on the side, you might want to update details here: https://www.internalpositioning.com/plugins/

File: openhabian-setup.sh [hunk 3]
```diff
@@ -1033,6 +1033,103 @@ To continue your integration in openHAB 2, please follow the instructions under:
...
+  FIND_SYSTEMCTL=/etc/systemd/system/findserver.service
+  FIND_DEFAULT=/etc/default/findserver
+  FIND_DSTDIR=/var/lib/findserver
+  MOSQUITTO_PASSWD=/etc/mosquitto/passwd
+  FIND_TMP=/tmp/find-latest.$$
+  CLIENT_TMP=/tmp/fingerprint-latest.$$
+
+  if [ ! -f ${MOSQUITTO_PASSWD} ]; then
+    whiptail --title "Cannot install FIND" --msgbox "FIND requires mosquitto to be installed first !" 10 60
+    return
```
Comment (ThomDietrich) [inline openhabian-setup.sh]: "return 1"

File: openhabian-setup.sh [hunk 4]
```diff
@@ -1033,6 +1033,103 @@ To continue your integration in openHAB 2, please follow the instructions under:
...
+  /bin/mkdir -p ${FIND_DSTDIR}
+
+  /usr/bin/wget -O ${FIND_TMP} ${FIND_SRC}
+  /usr/bin/wget -O ${CLIENT_TMP} ${CLIENT_SRC}
+
+  /usr/bin/unzip ${FIND_TMP} findserver -d /usr/sbin
+  /usr/bin/unzip ${FIND_TMP} fingerprint -d /usr/sbin
+  /usr/bin/unzip ${FIND_TMP} static\* -d ${FIND_DSTDIR}
+
+
```
Comment (ThomDietrich) [inline openhabian-setup.sh]: please remove superfluous empty lines

File: openhabian-setup.sh [hunk 5]
```diff
@@ -1033,6 +1033,103 @@ To continue your integration in openHAB 2, please follow the instructions under:
...
+ 15 80 3>&1 1>&2 2>&3)
+    echo "ADMIN=${FIND_ADMIN}" >>${FIND_DEFAULT}
+    echo "ADMIN_PASS=${FIND_ADMIN_PASS}" >>${FIND_DEFAULT}
+    cond_redirect /usr/bin/mosquitto_passwd -b ${MOSQUITTO_PASSWD} ${FIND_ADMIN} ${FIND_ADMIN_PASS}
+    if [ $? -ne 0 ]; then echo "FAILED"; exit 1; fi
+  fi
+
+  cond_redirect /bin/systemctl enable findserver.service
+  cond_redirect /bin/systemctl restart findserver.service
+  /bin/rm -f ${FIND_TMP} ${CLIENT_TMP}
```
Comment (ThomDietrich) [inline openhabian-setup.sh]: shouldn't we do that before starting the services?

File: openhabian-setup.sh [hunk 6]
```diff
@@ -1033,6 +1033,103 @@ To continue your integration in openHAB 2, please follow the instructions under:
...
+  if [ $? -eq 0 ]; then echo "OK"; else echo "FAILED"; exit 1; fi
+  
+  if [ -n "$INTERACTIVE" ]; then
+    if [ $FAILED -eq 0 ]; then
```
Comment (ThomDietrich) [inline openhabian-setup.sh]: this condition is never false

**Comment by mstormi** [comment]
what do you mean by the pointer to the FIND doc and "update details here" ?

**Comment by ThomDietrich** [comment]
I meant, there is a short paragraph on openHAB behind that link. Would be great if you could update it and maybe add a similar screenshot for Basic UI

**Comment by mstormi** [comment]
ok, I contacted the FIND authors to have them update their link to our tutorial on the community website.
Also uploaded [fixes](https://github.com/openhab/openhabian/pull/226/commits/282de9a458bab66892a5bf12c3bc479e035edbfe) based on your code comments above ... is there a "re-review" button ? (I'm still no git pro)

**Comment by ThomDietrich** [comment]
No there is not but your comment is the right way to make me aware.
Did we misunderstand each other? You've added the files but the idea was to of course now use those instead of having them inside the function...

**Comment by mstormi** [comment]
No we didn't misunderstand that's how I reworked it... ok I see that the link is my last post only shows the added includes files and not the changes to openhabian-setup.sh as well...  something must have gone wrong when I was rolling changes forth and back with git ...  seems those changes got lost... will get back to you when I figured that out.

**Comment by mstormi** [comment]
Phew, git can be confusing.... you never know for sure which branch it's actually working against...
anyway, here's my changes.
Glad I made an off-git copy before experimenting with git, otherwise it would have been lost.

**Comment by kubawolanin** [comment]
Hey @mstormi @ThomDietrich any updates on this? 😊 
Do you guys need any help to push it forward? 
Would love to try it out!

**Comment by mstormi** [comment]
Waiting for Thomas to re-review my code...
If you're eager to test, check out https://github.com/mstormi/openhabian/tree/patch-12

**Comment by ThomDietrich** [comment]
Hey guys, I did a quick check and I think everything looks fine now. I'm going to merge!

@mstormi one thing I noticed: You are checking for a local mqtt configuration, later you are asking the user for the hostname if his MQTT server (you should rename that to "MQTT broker"). Does that make sense? Feel free to open a new PR with a fix ;)

Thanks for the addition!!!

**Comment by mstormi** [comment]
Yes this is intentional. I'm unsure about the interface(s) that mosquitto listens to, also a user might have changed that after setting up mosquitto and before installing FIND.
That's why I want the user to enter it (again) and "localhost" is proposed as the default value.

**Pull request merged**